Plugins/Skinning/src/PositionBasedSkinning/PBS.cpp

```diff
@@ -0,0 +1,521 @@
+                {
+                    const unsigned int constraintIndex = groups[group][i];
+                    constraints[constraintIndex]->solvePositionConstraint(m_model);
+                }
+            }
+        }
+
+        iter++;
+    }
+}
+
+/* ----------------------
+  Time Integration
+ ----------------------- */
+
+void PBS::semiImplicitEuler(const Scalar h, ParticleData &pd)
+{
+    const unsigned int num_particles = pd.size();
+
+    for (auto i = 0; i < num_particles; i++)
+    {
+        if (pd.getMass(i) != 0.0)
+        {
+            pd.getVelocity(i) += pd.getAcceleration(i) * h;
+            pd.getPosition(i) += pd.getVelocity(i) * h;
+        }
+    }
+}
+
+void PBS::velocityUpdateFirstOrder(const Scalar h, ParticleData &pd)
+{
+    const unsigned int num_particles = pd.size();
+
+    for (auto i = 0; i < num_particles; i++)
+    {
+        if (pd.getMass(i) != 0.0)
+            pd.getVelocity(i) = (1.0f / h) * (pd.getPosition(i) - pd.getOldPosition(i));
+    }
+}
+
+//-----------------------------------------------------------------------------------------------------------
+
+void PBS::step(const Animation::Skeleton *skeleton, ParticleData &pd)
+{
+    TimeManager *tm = TimeManager::getCurrent ();
+    const Scalar h = tm->getTimeStepSize();
+
+    m_model.setSkeleton(skeleton);
+    clearAccelerations(pd);
+
+    semiImplicitEuler(h, pd);
+    positionConstraintProjection();
+    velocityUpdateFirstOrder(h, pd);
+
+    // compute new time
+    tm->setTime (tm->getTime () + h);
+}
+
+//----------------------------
+// Link to Surface Mesh
+//---------------------------------
+
+Scalar PBS::computeDeterminant( const Vector3 &vertex0, const Vector3 &vertex1,
+                            const Vector3 &vertex2, const Vector3 &vertex3 )
+{
+    Scalar det =  vertex1[2] * vertex2[1] * vertex3[0] - vertex0[2] * vertex2[1] * vertex3[0] -
+                vertex1[1] * vertex2[2] * vertex3[0] + vertex0[1] * vertex2[2] * vertex3[0] +
+
+                vertex0[2] * vertex1[1] * vertex3[0] - vertex0[1] * vertex1[2] * vertex3[0] -
+                vertex1[2] * vertex2[0] * vertex3[1] + vertex0[2] * vertex2[0] * vertex3[1] +
+
+                vertex1[0] * vertex2[2] * vertex3[1] - vertex0[0] * vertex2[2] * vertex3[1] -
+                vertex0[2] * vertex1[0] * vertex3[1] + vertex0[0] * vertex1[2] * vertex3[1] +
+
+                vertex1[1] * vertex2[0] * vertex3[2] - vertex0[1] * vertex2[0] * vertex3[2] -
+                vertex1[0] * vertex2[1] * vertex3[2] + vertex0[0] * vertex2[1] * vertex3[2] +
+
+                vertex0[1] * vertex1[0] * vertex3[2] - vertex0[0] * vertex1[1] * vertex3[2] -
+                vertex0[2] * vertex1[1] * vertex2[0] + vertex0[1] * vertex1[2] * vertex2[0] +
+
+                vertex0[2] * vertex1[0] * vertex2[1] - vertex0[0] * vertex1[2] * vertex2[1] -
+                vertex0[1] * vertex1[0] * vertex2[2] + vertex0[0] * vertex1[1] * vertex2[2];
+    return det;
```

Comment:
this is simply the signed volume of the tetrahedron, so this might be faster this way (from your code for volume constraint):
```c++
return (1.0f / 6.0f) * (vertex3 - vertex0).dot((vertex2 - vertex0).cross(vertex1 - vertex0))
```

Reply:
Please note that the volume constraint starts at line 97 in constraints.cpp, but this code here is used to compute the determinant for the barycentric coordinate for mapping between the surface and tetrahedral mesh.

Reply:
okay I did profile both and I confirm that they are more/less the same performance, but your suggestion has better readability, therefore I change it.